include/vcmi/spells/Caster.h

```diff
@@ -68,9 +69,9 @@ class DLL_LINKAGE Caster
 	virtual void spendMana(ServerCallback * server, const int32_t spellCost) const = 0;
 
 	virtual int32_t manaLimit() const = 0;
-	
-	///used to identify actual hero caster
-	virtual const CGHeroInstance * getHeroCaster() const = 0;
+
+	/// This caster ID should be used only in adventure spells (-1 if no adventure casting possible)
+	virtual ObjectInstanceID getAdventureCasterID() const = 0;
```

Comment:
but why?
Returning pointer is much more useful and clear than indirect usage with some index with further searching object though map. Those indexes are quite buggy btw.

Also, how did you test that you won't break functionality, which was recently introduced?

Reply:
@Nordsoft91 because caster should not know if there is a hero, unit or some other entity.

About testing - what should I test except adventure and battle spell casting of all spells (I did such testing without mods).

Reply:
I was not agree to #1938, just did not have rights to set RequestChanges) And this is my disagreement point BTW)

Reply:
Agree with you. See here: https://github.com/vcmi/vcmi/pull/2105#discussion_r1181228702

Reply:
@Nordsoft91 Will do a testing with all spells with all skill levels if you want.

Reply:
@Nordsoft91 if you absolutely think than that change should belongs to other PR - i can split and force-push.
But I do not think than such abomination like pointer inside caster interface should survive at least one release.

@IvanSavenko what do you think about getHeroCaster and my opinion strongly against it?

Reply:
This is new functionality - basically, town gate advmap object shall work. See mod attached here https://github.com/vcmi/vcmi/pull/1928
Also, rewardable object instructions on wiki (in case mod is obsolete)

Reply:
@Nordsoft91 will try. Maybe include something like it in unit test (but need think how to include)?

Reply:
@Nordsoft91 do you have a test map for that mod?

Reply:
In my opinion preference should be:
1) Pointer to class.
2) if not possible - then object ID in enum form.
3) if all else fails - object ID in integer form.

- If we want to avoid unnecessary "knowledge" then we can use for example "tag" class that potential caster must inherit from.
- Pointers are generally more robust & result in more clear code (since you don't need constant VLC lookups to find referenced object)
- Pointers are also more flexible, especially when using "tag" / interface class - since said interface can be implemented by any other class.

Reply:
@IvanSavenko but caster should not always be a hero, it can be an object (like Shipyard cast CreateBoat even without a hero involved). So, I think pointer to hero is not suitable for caster.

@Nordsoft91 your mod works, town portal can be cast by campfire, but console spams with ```Removal of object with spell casts is not supported!```. It is ok?

Reply:
@Nordsoft91 used this form 
```json
					{
						"spellCast": {
							"spell" : "townPortal",
							"schoolLevel" : 3
						},
						"removeObject": true
					}
```

Reply:
> I was not agree to https://github.com/vcmi/vcmi/pull/1928, just did not have rights to set RequestChanges) And this is my disagreement point BTW)

Sorry, but I didn't understood that you was not agree.
Don't care if feature works and users are happy, but your approach is very questionable


Your approach won't work for "virtual" casters, you obligate using objects from map only. 
Closest example  - town building.

Reply:
> @IvanSavenko but caster should not always be a hero, it can be an object (like Shipyard cast CreateBoat even without a hero involved). So, I think pointer to hero is not suitable for caster.

This is why I said about "tag class". Create an interface class, let's call it CasterBearer and make Shipyard / CGHeroInstance / etc inherit from it. With such class:
- it is impossible to misunderstand object ID / implicitly convert it into other ID class somewhere (we have a LOT of such type conversions)
- compiler will perform all checks for you while object ID can point to another object that can not be a caster
- it is possible to implement caster that won't actually be present on map. Town building is a great example here.
- developer also can quickly find which classes are valid as casters and which are not

Reply:
@Nordsoft91 why? I thought town building will set an ID of parent town or of visiting/garrison hero. 

@IvanSavenko I think any classes which derived from spells::Caster should be able to cast a spell. But actually we can derive something like AdventureCaster, if you wish. 

Anyway, I think IDs is better than "tag classes", because IDs can be passed around as int and much easier to check.

@Nordsoft91 Another case why I was against getHeroCaster, because you reworked all battle checks to use getHeroCaster (example is #1964). I think getCasterUnitID should remain for battle casters. I can preserve getHeroCaster temporarily for adventure casters, if desired.

Reply:
>  I thought town building will set an ID of parent town or of visiting/garrison hero.

So town building cannot be casters in your solution. Only towns. 

> IDs can be passed around as int and much easier to check.

what do you mean easier to check? It's like same saying why not using reinterpret_cast<chat*> for everything.
vice-versa, we should avoid using ids, where possible. How you will debug ids? how you will identify whether it's valid or not? I remember several crashes with Ids, they are very painful to debug and fix

>  I think getCasterUnitID should remain for battle casters

What for? Can you give me an example of use case?

Reply:
Adventure casters can be used with getHeroCaster, and there is a purpose of it. But getCasterUnitID was used  to return unitID or -1 (it which case we should assume than cast was perfrormed by hero or by mechanics like moat setup). You interfere this semantics in #1928, when you made CGHeroInstance return its hero ID instead of -1, and it introduced a several bugs which was fixed hackish way in #1964,  introducing getHeroCaster inside battle classes, it is what I want to avoid most (even more than CGHeroInstance* inside spells::Caster). In this PR I want to separate semantics again and make getBattleCasterID (which will work as getCasterUnitID pre-#1928) and other way to get adventure caster only.

Reply:
Technically going from abstraction to exact implementation shows that initial abstraction leaks. Why do we need to know who casted a spell?

Reply:
> But getCasterUnitID was used only in battle to return unitID or -1 (it which case we should assume than cast was perfrormed by hero or by mechanics like moat setup)

Literally, algorithm is following
```
if caster is not hero
  get unit id
```

before done implicitly 
```
if(getCasterUnitID() >= 0) //don't understand what is that? checking that caster is hero, obviously
  ...
```

now explicitly
```
if(getHeroCaster() == nullptr)
  ...
```

Reply:
If this method is such a mess then why have it to begin with?
Perhaps remove it / replace with something else completely?

What are exact use-cases for this method?

Reply:
> if caster is not hero
>  get unit id
Exactly. One has to figure out that there are only two types of caster unit and hero so if it is not hero - it is unit. And it breaks as soon as we have our initial assumption not true anymore, like moats casting or something.
So direct check if caster is unit then get unit id is way better

Reply:
@IvanSavenko @nullkiller caster is not always a hero or unit. There is a case when caster is just a caster.

Reply:
Like most setup)

Reply:
So getCasterType() should work? I am afraid that people will think that getBattleCasterID() always return unit ID while its name is generic and it should be able to return something else. Also as I said hero is both battle and adventure caster so not clear if we should return something from both methods for hero.

Reply:
> @IvanSavenko @nullkiller caster is not always a hero or unit. There is a case when caster is just a caster.

Again, what is the use-case of this method? Perhaps we can remove it or replace with something else?

For example, `bool isHeroCaster()` or `bool hasHero()` or something like that

Reply:
1. In some cases we should know exactly what unit cast a spell (eg. Bind) 
2. Check whether is unit casts a spell or a non-unit (mostly for logs). 
3. Get unit ID for logs. 

Maybe more, this method is popular.

Reply:
- was there any need to change this method? If no - then I suggest revert at least until acceptable solution is found.
- is there any actual improvement in using object ID's over pointers? "Abstraction leaking" is a good argument, but replacing pointer with ID is not a fix for abstraction leak, but simply making it less obvious and more bug-prone.

Reply:
@IvanSavenko Ok, I will move this commit into a separate PR.

But I fundamentally disagree with this approach. This is my opinion below:

1. Anything in include/vcmi should not use concrete types, not a single concrete type, except IDs and only when we should not convert an IDs to another IDs. So, no type knowledge at all except types already presented in include/vcmi.
2. #1928 should remain one and only one example of circumventing this principle, all other PRs should be blocked if they circumvent this principle. And after merge, this circumventing was lead to other problems (like #1964 and introduction of getHeroCaster inside battle types).

I want to fix it myself, but I do not understand why all other developers disagree.

Reply:
Your controversal commit is moved outside of this PR.